halo2_proofs/src/plonk/prover.rs

```diff
@@ -436,6 +436,30 @@ where
         })
         .collect::<Result<Vec<_>, _>>()?;
 
+    let shuffles = instance
+        .iter()
+        .zip(advice.iter())
+        .map(|(instance, advice)| -> Vec<_> {
+            // Compress expressions for each shuffle
+            pk.vk
+                .cs
+                .shuffles
+                .iter()
+                .map(|shuffle| {
+                    shuffle.compress(
+                        pk,
+                        params,
+                        domain,
+                        theta,
+                        &advice.advice_polys,
+                        &pk.fixed_values,
+                        &instance.instance_values,
+                        &challenges,
+                    )
+                })
+                .collect::<Vec<_>>()
+        });
+
```

Comment:
It seems we could merge this part to the below one to save memory (we can discard the compressed values after committing to product.